NetworkConfigurations/TrafficControlAndPolicingSubCategories/TrafficShaping.md

```diff
@@ -0,0 +1,506 @@
+
+When creating a new qdisc to attach to a band, use the corresponding class ID as the parent reference. For example, to attach RED to the first band and TBF to the second one:
+
+```
+tc qdisc add dev swp23 root handle 10: htb
+tc class add dev swp23 parent 10: classid 10:1 htb rate 100mbit
+
+tc qdisc replace dev swp23 parent 10:1 handle 101: red limit 2M avpkt 1000 probability 0.1 min 500K max 1.5M
+```
+
+This configures RED (Random Early Detection) qdisc with specified parameters and attaches it to the first band. Similarly, TBF (Token Bucket Filter) qdisc can be attached to other bands as required.
+
+```
+RED: set bandwidth to 10Mbit
+```
+
+The output confirms that the bandwidth for RED qdisc has been successfully set to 10Mbit.
+
+```
+tc class show dev swp23
+```
+
+```
+class htb 10:1 root leaf 101: prio 0 rate 100Mbit ceil 100Mbit burst 1600b cburst 1600b
+class red 101:1 parent 101:
+```
+
+The output of `tc class show` command displays the configured classes, including the htb (Hierarchy Token Bucket) class with rate and ceiling set to 100Mbit, and the RED class.
+
+```
+tc -s qdisc show dev swp23
+```
+
+```
+qdisc htb 10: root refcnt 2 r2q 10 default 0 direct_packets_stat 0 direct_qlen 1000
+ Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
+ backlog 0b 0p requeues 0
+qdisc red 101: parent 10:1 limit 2Mb min 500Kb max 1536Kb
+ Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
+ backlog 0b 0p requeues 0
+  marked 0 early 0 pdrop 0 other 0
+qdisc clsact ffff: parent ffff:fff1
+ Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
+ backlog 0b 0p requeues 0
+```
+
+The output of `tc -s qdisc show` command provides detailed information about the configured qdiscs, including statistics and settings for each. In this case, it shows the HTB and RED qdiscs configured on the device `swp23`.
+
+Verfication using `iperf` utility:
+
+```
+iperf -c 127.0.0.1 -t 60 -b 8M
+```
+
+Please make sure to run your iperf server using `iperf -s` command.
+
+Output-
+
+```
+------------------------------------------------------------
+Client connecting to 127.0.0.1, TCP port 5001
+TCP window size: 2.50 MByte (default)
+------------------------------------------------------------
+[  3] local 127.0.0.1 port 52146 connected with 127.0.0.1 port 5001
+[ ID] Interval       Transfer     Bandwidth
+[  3]  0.0-60.0 sec  57.2 MBytes  8.00 Mbits/sec
+```
+
+Given that the configuration has a bandwidth limit of 10 Mbit/s, setting the iperf bandwidth slightly below this limit to observe the effect of traffic shaping is appropriate. In this case, the achieved bandwidth aligns with the specified bandwidth, indicating that the traffic shaping configuration is working as expected, effectively limiting the bandwidth to the configured rate of 8 Mbps.
```

Comment:
Bandwidth is limited to 10Mb/s. If we send 8Mb/s we should expect 8Mb/s, which we see. Shouldn't the real bandwidth limiting be observed by sending more than 10Mb/s through iperf and observe 10Mb/s?

I would keep the 8Mb/s example to show traffic is not shaped. But add sending something like 15Mb/s and observe 10Mb/s.